lahja/base.py

```diff
@@ -200,3 +193,86 @@ def __reduce__(self) -> None:  # type: ignore
+
+    def is_remote_subscribed_to(
+        self, remote_endpoint: str, event_type: Type[BaseEvent]
+    ) -> bool:
+        """
+        Return ``True`` if the specified remote endpoint is subscribed to the specified event type
+        from this endpoint. Otherwise return ``False``.
+        """
+        for endpoint, subscriptions in self.get_connected_endpoints_and_subscriptions():
+            if endpoint != remote_endpoint:
+                continue
+
+            for subscribed_event_type in subscriptions:
+                if subscribed_event_type == event_type:
+                    return True
+
+        return False
+
+    def is_any_remote_subscribed_to(self, event_type: Type[BaseEvent]) -> bool:
+        """
+        Return ``True`` if at least one of the connected remote endpoints is subscribed to the
+        specified event type from this endpoint. Otherwise return ``False``.
+        """
+        for endpoint, subscriptions in self.get_connected_endpoints_and_subscriptions():
+            for subscribed_event_type in subscriptions:
+                if subscribed_event_type == event_type:
+                    return True
+
+        return False
+
+    def are_all_remotes_subscribed_to(self, event_type: Type[BaseEvent]) -> bool:
+        """
+        Return ``True`` if every connected remote endpoint is subscribed to the specified event
+        type from this endpoint. Otherwise return ``False``.
+        """
+        for endpoint, subscriptions in self.get_connected_endpoints_and_subscriptions():
+            if event_type not in subscriptions:
+                return False
+
+        return True
+
+    async def wait_until_remote_subscribed_to(
+        self, remote_endpoint: str, event: Type[BaseEvent]
+    ) -> None:
+        """
+        Block until the specified remote endpoint has subscribed to the specified event type
+        from this endpoint.
+        """
+
+        if self.is_remote_subscribed_to(remote_endpoint, event):
+            return
+
+        async for _ in self.stream(RemoteSubscriptionChanged):  # noqa: F841
+            if self.is_remote_subscribed_to(remote_endpoint, event):
+                return
+
+    async def wait_until_any_remote_subscribed_to(self, event: Type[BaseEvent]) -> None:
+        """
+        Block until any other remote endpoint has subscribed to the specified event type
+        from this endpoint.
+        """
+
+        if self.is_any_remote_subscribed_to(event):
+            return
+
+        async for _ in self.stream(RemoteSubscriptionChanged):  # noqa: F841
+            if self.is_any_remote_subscribed_to(event):
+                return
+
+    async def wait_until_all_remotes_subscribed_to(
+        self, event: Type[BaseEvent]
+    ) -> None:
+        """
+        Block until all currently connected remote endpoints are subscribed to the specified
+        event type from this endpoint.
+        """
+
+        if self.are_all_remotes_subscribed_to(event):
+            return
+
```

Comment:
I think there's a race-condition if we receive `RemoteSubscriptionChanged` right here. It'll be dropped and then this method will block until the next `RemoteSubscriptionChanged` comes in, which might take a while.

Reply:
But how would that happen? We aren't using multiple threads (are we?) and a context switch can only happen at the next `await`, no?

Reply:
You're right. I looked more carefully and there's *almost* a race-condition because `stream` blocks on notifying before it starts reading from the queue, but it adds the queue first so you're right that there are no switch points between the check above this and waiting for the message.

Reply:
I've been tinkering in another branch and have a deadlock that I'm trying to unravel.  Makes me want a solid test suite around these APIs because it seem to be quite easy to get into a situation where we can deadlock like this.

Reply:
>That's assuming that we only switch when the co-routine actually blocks, which I think is the default asyncio behavior

Correct. I just refreshed my memory ([see this](https://stackoverflow.com/questions/54860285/how-can-i-prevent-context-switching-when-calling-an-async-function/54863862#54863862) and [that](https://djangostars.com/blog/asynchronous-programming-in-python-asyncio/)) and this is how it is supposed to work.

>In the unlikely event that uvloop could switch away at any `await` call then I think there's still a race-condition, because there's an implicit await when we loop over `self.stream`

Correct, but as far as I understand things (and it really makes my head spin!), this should not be the case.

@pipermerriam 

>I've been tinkering in another branch and have a deadlock that I'm trying to unravel

Do you have a link for the curious?

At the risk of pointing out the obvious, you do use `PYTHONASYNCIODEBUG=1` for debugging these, right? I've found that to be a real life safer.

Reply:
It's trio stuff and it's in #56 and nowhere near ready for anything beyond curious observers.  Still major structural changes happening as I re-shape the core `asyncio` library implementation to a form that's close enough to the future `trio` implementation that they can properly interop without gross hacks.